core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -35,92 +35,109 @@ oppia.factory('NumberWithUnitsObjectFactory', [
       this.type = type;
       this.real = real;
       this.fraction = fractionObj;
-      this.units = unitsObj;
+      this.units = unitsObj.units;
     };
 
     NumberWithUnits.prototype.toString = function() {
       var numberWithUnitsString = '';
-      if (this.units.toString().includes('$')) {
-        this.units.units = this.units.toString().replace('$', '');
+      var unitsString = UnitsObjectFactory.fromList(this.units).toString();
+      if (unitsString.includes('$')) {
+        unitsString = unitsString.replace('$', '');
         numberWithUnitsString += '$' + ' ';
       }
-      if (this.units.toString().includes('Rs')) {
-        this.units.units = this.units.toString().replace('Rs', '');
+      if (unitsString.includes('Rs')) {
+        unitsString = unitsString.replace('Rs', '');
         numberWithUnitsString += 'Rs' + ' ';
       }
-      if (this.units.toString().includes('₹')) {
-        this.units.units = this.units.toString().replace('₹', '');
+      if (unitsString.includes('₹')) {
+        unitsString = unitsString.replace('₹', '');
         numberWithUnitsString += '₹' + ' ';
       }
 
       if (this.type === 'real') {
         numberWithUnitsString += this.real + ' ';
-      } else {
+      } else if (this.type === 'fraction') {
         numberWithUnitsString += this.fraction.toString() + ' ';
       }
-      numberWithUnitsString += this.units.toString();
+      numberWithUnitsString += unitsString.trim();
       numberWithUnitsString = numberWithUnitsString.trim();
 
       return numberWithUnitsString;
     };
 
+    NumberWithUnits.prototype.toDict = function() {
+      return {
+        type: this.type,
+        real: this.real,
+        fraction: this.fraction.toDict(),
+        units: this.units.toDict().units
```

Comment:
`this.units` is already equal to units list (and not UnitsObjectFactory object). Maybe add a few frontend tests for this method?

Reply:
Done, modified!